Extend impl_backward to be usable with torch.library operators

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #106817
* #106800
* #106799

- impl_save_for_backward/impl_backward only work for functional,
non-view schemas. We validate this.
- impl_save_for_backward/impl_backward raise if there already exists an
autograd implementation from torch.library / TORCH_LIBRARY.
- Operators constructed via custom_op receive an "autograd indirection
kernel". The "autograd indirection kernel" automatically pulls the
constructed autograd kernel out of a dict. When
impl_save_for_backward/impl_backward get used with torch.library
operators, we also register the "autograd indirection kernel" so we can
reuse the logic.

Test Plan:
- new tests

zou3519 committed:
Extend impl_backward to be usable with torch.library operators

- impl_save_for_backward/impl_backward only work for functional,
non-view schemas. We validate this.
- impl_save_for_backward/impl_backward raise if there already exists an
autograd implementation from torch.library / TORCH_LIBRARY.
- Operators constructed via custom_op receive an "autograd indirection
kernel". The "autograd indirection kernel" automatically pulls the
constructed autograd kernel out of a dict. When
impl_save_for_backward/impl_backward get used with torch.library
operators, we also register the "autograd indirection kernel" so we can
reuse the logic.

Test Plan:
- new tests

[ghstack-poisoned]

soulitzer commented on torch/_custom_op/autograd.py:
```diff
@@ -102,7 +103,7 @@ def forward(ctx, *flat_args):
             args = pytree.tree_unflatten(list(flat_args), spec)
             with torch._C._AutoDispatchBelowAutograd():
-                output = forward_op(*args)
+                output = op_overload(*args)
```
This makes it so that instead of calling the special logic in` __call__` of custom_op, we now call into `torch.ops*` directly. How much should we worry about what the comment says in `__call__` about directly calling into `torch.ops*` doing bad things, or are we saying with the new updated API we don't really care about users calling into the op through`torch.ops.*` anymore?

zou3519 replied:
> are we saying with the new updated API we don't really care about users calling into the op throughtorch.ops.* anymore?

I'm going to delete `CustomOp.__call__` at some point. We're telling people to use the torch.ops.blah now. There are some pitfalls with it IF you deregister the operator, but no one actually does that in practice.

soulitzer replied:
I see, thanks!

soulitzer reviewed (approved):
Nice! 
just had a question about calling into ops through custom_op vs through `torch.ops.*`

zou3519 commented on torch/_custom_op/impl.py:
```diff
@@ -963,7 +1025,10 @@ def custom_op_from_existing(op):
+    schema_str = str(op._schema)
+    # CustomOp expects the schema string without the namespace
+    schema_str = schema_str.split("::")[-1]
+    schema = FunctionSchema.parse(schema_str)
```
Nope haha

zou3519 committed:
Update on "Extend impl_backward to be usable with torch.library operators"


- impl_save_for_backward/impl_backward only work for functional,
non-view schemas. We validate this.
- impl_save_for_backward/impl_backward raise if there already exists an
autograd implementation from torch.library / TORCH_LIBRARY.
- Operators constructed via custom_op receive an "autograd indirection
kernel". The "autograd indirection kernel" automatically pulls the
constructed autograd kernel out of a dict. When
impl_save_for_backward/impl_backward get used with torch.library
operators, we also register the "autograd indirection kernel" so we can
reuse the logic.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "unrelated failures"
